### PR TITLE
optionally allow WPS

### PR DESCRIPTION
**Pull request by d-a-v**
* Disable WPS at compile time
* At the price of 4K in heap, allow WPS by using a `boards.txt.py` option
* documentation in FAQ

ref: #4779

### CODE REVIEW

File: tools/sdk/include/user_interface.h [hunk 1]
```diff
@@ -575,9 +575,16 @@ enum wps_cb_status {
...
 bool wifi_wps_disable(void);
 bool wifi_wps_start(void);
+#else
+bool WPS_is_unavailable_in_this_configuration__Please_check_FAQ_or_board_generator_tool ();
+#define wifi_wps_enable(...) WPS_is_unavailable_in_this_configuration__Please_check_FAQ_or_board_generator_tool()
+#define wifi_wps_disable() WPS_is_unavailable_in_this_configuration__Please_check_FAQ_or_board_generator_tool()
+#define wifi_wps_start() WPS_is_unavailable_in_this_configuration__Please_check_FAQ_or_board_generator_tool()
+#endif
 
 typedef void (*wps_st_cb_t)(int status);
```
Comment (devyte) [inline tools/sdk/include/user_interface.h]: Should these 2 lines be moved into the #ifdef ?

File: cores/esp8266/core_esp8266_main.cpp [hunk 2]
```diff
@@ -156,11 +156,25 @@ void init_done() {
  * Peripherals (except for SPI0 and UART0) are not initialized.
  * This function does not return.
  */
+
```
Comment (devyte) [inline cores/esp8266/core_esp8266_main.cpp]: Please add a descriptive comment about what the #define means (explain it's either WPS or stack optimization, but not both), or at least reference the issue.

File: doc/faq/readme.rst [hunk 3]
```diff
@@ -40,6 +40,22 @@ entering an issue report, please perform initial troubleshooting.
...
+Using ``*printf()`` with floats is enabled by default.  Some KBs of flash can
+be saved by using the option ``--nofloat`` with the boards generator:
+
+``./tools/boards.txt.py --nofloat --allgen``
+
+Why can't I use WPS ?
+~~~~~~~~~~~~~~~~~~~~~
+
+WPS is disabled by default, this offers an extra 4KB in ram/heap.  To enable
+WPS, use this boards generator option:
```
Comment (devyte) [inline doc/faq/readme.rst]: Please state explicitly that enabling WPS removes the 4KB heap optimization.

File: libraries/ESP8266WiFi/src/ESP8266WiFiSTA.cpp [hunk 4]
```diff
@@ -571,6 +571,8 @@ int32_t ESP8266WiFiSTAClass::RSSI(void) {
 // -------------------------------------------------- STA remote configure -----------------------------------------------
 // -----------------------------------------------------------------------------------------------------------------------
 
+#ifdef NO_EXTRA_4K_HEAP
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/ESP8266WiFiSTA.cpp]: Please add a descriptive comment about what the #define means, or at least reference the issue.

File: libraries/ESP8266WiFi/src/ESP8266WiFiSTA.h [hunk 5]
```diff
@@ -92,7 +93,13 @@ class ESP8266WiFiSTAClass {
 
     public:
 
+#ifdef NO_EXTRA_4K_HEAP
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/ESP8266WiFiSTA.h]: Please add a descriptive comment about what the #define means, or at least reference the issue.

File: tools/sdk/include/user_interface.h [hunk 6]
```diff
@@ -575,9 +575,16 @@ enum wps_cb_status {
     WPS_CB_ST_UNK,
 };
 
+#ifdef NO_EXTRA_4K_HEAP
```
Comment (devyte) [inline tools/sdk/include/user_interface.h]: Please add a descriptive comment about what the #define means, or at least reference the issue.

@igrr In this case, should the change be moved into a patch that gets applied over the original user_interface.h file? I'm thinking of migration to future sdk versions, and what will happen with our own changes to this file.

**Comment by devyte** [comment]
@d-a-v this is pretty good. My sole requests is to add comments to explain what it means. I'm thinking of the future when we come back and look at this, to avoid head scratching/WTFs and so on.

File: cores/esp8266/core_esp8266_main.cpp (hunk 2 above)
Comment (d-a-v) [reply cores/esp8266/core_esp8266_main.cpp]: Would this (above, below) be a clear enough explanation with its references ?

File: doc/faq/readme.rst [hunk 7]
```diff
@@ -40,6 +40,22 @@ entering an issue report, please perform initial troubleshooting.
...
+Using ``*printf()`` with floats is enabled by default.  Some KBs of flash can
+be saved by using the option ``--nofloat`` with the boards generator:
+
+``./tools/boards.txt.py --nofloat --allgen``
+
+Why can't I use WPS ?
+~~~~~~~~~~~~~~~~~~~~~
+
+WPS is disabled by default, this offers an extra 4KB in ram/heap.  To enable
+WPS (and loose 4KB of useable ram), use this boards generator option:
```
Comment (devyte) [inline doc/faq/readme.rst]: loose -> lose

File: cores/esp8266/core_esp8266_main.cpp [hunk 8]
```diff
@@ -156,11 +156,62 @@ void init_done() {
  * Peripherals (except for SPI0 and UART0) are not initialized.
  * This function does not return.
  */
+/*
+   A bit of explanation for this entry point:
+
+   SYS is the RTOS task/context used by the upperlying system to run its
```
Comment (devyte) [inline cores/esp8266/core_esp8266_main.cpp]: RTOS task/context -> SDK task context

There is confusion about RTOS and NONOS, I'd prefer to not add to that.

**Pull request merged**